server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/DesiredBalanceReconciler.java

```diff
@@ -449,4 +461,51 @@ private Decision decideCanForceAllocateForVacate(ShardRouting shardRouting, Rout
         assert target != null : "Target node is not found";
         return allocation.deciders().canForceAllocateDuringReplace(shardRouting, target, allocation);
     }
+
+    private OrderedNodesShardsIterator createOrderedNodesShardsIterator() {
+        return new OrderedNodesShardsIterator();
+    }
+
+    private class OrderedNodesShardsIterator implements Iterator<ShardRouting> {
+
+        private final ArrayDeque<NodeAndShardIterator> queue;
+
+        OrderedNodesShardsIterator() {
+            this.queue = new ArrayDeque<>(routingNodes.size());
+            for (var nodeId : moveOrdering.sort(routingNodes.getAllNodeIds())) {
+                var node = routingNodes.node(nodeId);
+                if (node.size() > 0) {
+                    this.queue.add(new NodeAndShardIterator(nodeId, Iterators.forArray(node.copyShards())));
+                }
+            }
+        }
+
+        @Override
+        public boolean hasNext() {
+            return queue.isEmpty() == false;
+        }
+
+        @Override
+        public ShardRouting next() {
+            if (queue.isEmpty()) {
+                throw new NoSuchElementException();
+            }
+            var entry = queue.peek();
+            assert entry.iterator.hasNext();
+            final var nextShard = entry.iterator.next();
+            if (entry.iterator.hasNext() == false) {
+                queue.poll();
+            }
+            return nextShard;
+        }
+
+        public void dePrioritizeNode(String nodeId) {
```

Comment:
nit: IMO we should update `moveOrdering` here rather than relying on the caller calling both `dePrioritizeNode` and `recordAllocation`.